src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -240,9 +258,81 @@ public void Dispose()
             {
                 if (this.toRelease != null)
                 {
-                    this.toRelease.semaphore.Release();
+                    this.toRelease.Release();
+                }
+            }
+        }
+
+        private static void CancellationHandler(object? state)
+        {
+            var waiterInfo = (WaiterInfo)state!;
+
+            // The party that manages to complete or cancel the task is responsible to remove it from the queue.
+            if (waiterInfo.Trigger.TrySetCanceled(waiterInfo.CancellationToken.IsCancellationRequested ? waiterInfo.CancellationToken : new CancellationToken(true)))
+            {
+                // If the node is in the queue, remove it.
+                // It might not have been added yet if cancellation was already requested by the time we called Register.
+                lock (waiterInfo.Owner.syncObject)
+                {
+                    if (waiterInfo.Owner.waiters.Find(waiterInfo) is { } node)
```

Comment:
Does this potentially create an O(N^2) algorithm if lots of waiters get canceled?

Reply:
I suppose it does. An alternative would be to store the `LinkedListNode` itself so we don't have to `Find` it. But since I recycle `LinkedListNode` I'm concerned about removing it in this handler even if the node has since been repurposed. Any suggestions?

Reply:
I've eliminated the O(n) search, and I think we're still thread-safe for recycling.